storage.go

```diff
@@ -3,34 +3,162 @@ package minio
 import (
 	"context"
 	"io"
+	"strings"
 
+	"github.com/minio/minio-go/v7"
+
+	ps "github.com/beyondstorage/go-storage/v4/pairs"
+	"github.com/beyondstorage/go-storage/v4/pkg/iowrap"
+	"github.com/beyondstorage/go-storage/v4/services"
 	. "github.com/beyondstorage/go-storage/v4/types"
 )
 
+const defaultListObjectBufferSize = 100
+
 func (s *Storage) create(path string, opt pairStorageCreate) (o *Object) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		if !s.features.VirtualDir {
+			return
+		}
+		rp += "/"
+		o = s.newObject(true)
+		o.Mode = ModeDir
+	} else {
+		o = s.newObject(false)
+		o.Mode = ModeRead
+	}
+	o.ID = rp
+	o.Path = path
+	return o
 }
 
 func (s *Storage) delete(ctx context.Context, path string, opt pairStorageDelete) (err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		if !s.features.VirtualDir {
+			err = services.PairUnsupportedError{Pair: ps.WithObjectMode(opt.ObjectMode)}
+			return
+		}
+		rp += "/"
+	}
+	err = s.client.RemoveObject(ctx, s.bucket, rp, minio.RemoveObjectOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
 }
 
 func (s *Storage) list(ctx context.Context, path string, opt pairStorageList) (oi *ObjectIterator, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	options := minio.ListObjectsOptions{WithMetadata: false}
+	switch {
+	case opt.ListMode.IsDir():
+		options.Recursive = false
+		if !strings.HasSuffix(rp, "/") {
+			rp += "/"
+		}
+	case opt.ListMode.IsPrefix():
+		options.Recursive = true
+	default:
+		return nil, services.ListModeInvalidError{Actual: opt.ListMode}
+	}
+	options.Prefix = rp
+
+	input := &objectPageStatus{
+		bufferSize: defaultListObjectBufferSize,
+		options:    options,
+	}
+	return NewObjectIterator(ctx, s.nextObjectPage, input), nil
 }
 
 func (s *Storage) metadata(opt pairStorageMetadata) (meta *StorageMeta) {
-	panic("not implemented")
+	meta = NewStorageMeta()
+	meta.Name = s.bucket
+	meta.WorkDir = s.workDir
+	return meta
+}
+
+func (s *Storage) nextObjectPage(ctx context.Context, page *ObjectPage) error {
+	input := page.Status.(*objectPageStatus)
+	if input.objChan == nil {
+		input.objChan = s.client.ListObjects(ctx, s.bucket, input.options)
+	}
+
+	for i := 0; i < input.bufferSize; i++ {
+		v, ok := <-input.objChan
+		if !ok {
+			return IterateDone
+		}
+		if v.Err != nil {
+			return v.Err
+		}
+		o, err := s.formatFileObject(v)
+		if err != nil && err != services.ErrObjectNotExist {
```

Comment:
```suggestion
		if err != nil && !errors.Is(err, services.ErrObjectNotExist) {
```

Reply:
Why we need to check `ErrObjectNotExist` in `formatFileObject`?

Reply:
It's because `ErrObjectNotExist` was returned in `formatFileObject`. This is another question, why check `v.Etag` and return `ErrObjectNotExist` when it's etag is blank?

Reply:
minio `list` api returns dir without Etag. Here is an example. So I just pass it.

```
Object:
{39e6d200da076f029669ddc6a5b24f95 client.go 2021-07-02 02:13:49.417 +0000 UTC 404  0001-01-01 00:00:00 +0000 UTC map[] map[] map[] {minio 02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4} [] STANDARD false false  <nil>}
Dir:
{ go/ 0001-01-01 00:00:00 +0000 UTC 0  0001-01-01 00:00:00 +0000 UTC map[] map[] map[] { } []  false false  <nil>}
```

Reply:
It's better to format this object as a dir object instead of returning `ErrObjectNotExist`.

You can set the Object mode to `ModeDir` instead.